File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        """
        stack=[]
        dict1={
            '(':')',
            '{':'}',
            '[':']'
        }
        for i in s:
            if i in dict1.keys():
                stack.append(i)
            else:
                if stack==[]:
                    return False
                else:
                    a=stack.pop()
                    if i != dict1[a]:
                        return False
        
        return stack==[]
        """

        stack=[]
        dict1={
            '(':')',
            '{':'}',
            '[':']'
        }

        for x in s:
            if x in dict1.keys():
                stack.append(x)
            else:
                if stack==[]:
                    return False
                else:
                    a=stack.pop()
                    if x!=dict1[a]:
                        return False
        return stack==[]



















                

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        
        HashMap<Character, Character> map = new HashMap<>();
        map.put('}','{');
        map.put(']','[');
        map.put(')','(');
        
        Stack<Character> stack = new Stack<>();
        
        for (int i=0; i<s.length();i++) {
            if (map.containsValue(s.charAt(i))) {
                stack.push(s.charAt(i));
            } else {
                if (!stack.isEmpty()) {
                    char a = stack.pop();
                    if (a != map.get(s.charAt(i))) {
                        return false;
                }   
                } else {
                    return false;
                }
                    
            }
        }
                if (stack.isEmpty()) {
                    return true;
                } else {
                    return false;
                }
        
    }
}